crates/matrix-sdk/src/authentication/oidc/mod.rs

```diff
@@ -644,6 +644,46 @@ impl Oidc {
         Ok(Some(url))
     }
 
+    /// Get the account management URL where the user can manage their
+    /// identity-related settings.
+    ///
+    /// # Arguments
+    ///
+    /// * `action` - An optional action that wants to be performed by the user
+    ///   when they open the URL. The list of supported actions by the account
+    ///   management URL can be found in the [`VerifiedProviderMetadata`], or
+    ///   directly with [`Oidc::account_management_actions_supported()`].
+    ///
+    /// Returns `Ok(None)` if the URL was not found. Returns an error if the
+    /// request to get the provider metadata fails or the URL could not be
+    /// parsed.
+    ///
+    /// This method will cache the URL for a while, if the cache is not
+    /// populated it will internally call
+    /// [`Oidc::fetch_account_management_url()`] and cache the resulting URI
+    /// before returning it.
+    pub async fn account_management_url(
+        &self,
+        action: Option<AccountManagementActionFull>,
+    ) -> Result<Option<Url>, OidcError> {
+        let mut cache = self.client.inner.caches.account_management_url.lock().await;
+
+        let serialized_action = serde_json::to_string(&action)
+            .expect("We should be able to serialize the account management acction");
+
+        if let Some(url) = cache.get(&serialized_action) {
+            Ok(Some(url))
+        } else {
+            let maybe_url = self.fetch_account_management_url(action).await?;
+
+            if let Some(url) = &maybe_url {
+                cache.insert(serialized_action, url.clone());
+            }
+
+            Ok(maybe_url)
+        }
```

Comment:
If I am understanding this correctly, we are caching one URL per action? It seems wasteful to make one request per action since the base URL is shared by all actions.

We could just cache the base URL and call `build_account_management_url()` to get the full URL after that.

Reply:
I guess you're right, caching the whole metadata seems more reasonable: https://github.com/matrix-org/matrix-rust-sdk/pull/4663/commits/9575c6058154dc4d41173c1000b295eb490e876d.

Reply:
Good, I was wondering if that would make sense after #4550 as well.